### PR TITLE
Lazy big int array

### PR DESCRIPTION
**Pull request by jhartman**
Upgrades bobo for lazy big int array implementation of count arrays

### CODE REVIEW

File: sensei-core/src/test/java/com/senseidb/search/req/mapred/FacetCountsMapReduce.java [hunk 1]
```diff
@@ -28,11 +28,11 @@ public void init(JSONObject params) {
     if (!facetCountAccessor.areFacetCountsPresent()) {
       return null;
     }
-    int[] countDistribution = facetCountAccessor.getFacetCollector(column).getCountDistribution();
+    BigSegmentedArray countDistribution = facetCountAccessor.getFacetCollector(column).getCountDistribution();
     TermValueList termValueList = accessor.getTermValueList(column);
-    HashMap<String, IntContainer> ret = new HashMap<String, IntContainer>(countDistribution.length);
-    for (int i = 0; i < countDistribution.length; i++) {
-      ret.put(termValueList.get(i), new IntContainer(countDistribution[i]));
+    HashMap<String, IntContainer> ret = new HashMap<String, IntContainer>(countDistribution.size());
```
Comment (chtyim) [inline sensei-core/src/test/java/com/senseidb/search/req/mapred/FacetCountsMapReduce.java:52]: It's not incorrect, but it's always good practice to use interface (Map in this case) then specific implementation for variable type.

Comment (jhartman) [inline sensei-core/src/test/java/com/senseidb/search/req/mapred/FacetCountsMapReduce.java:52]: Terrence, unfortunately the design of the map-reduce API makes this impossible. The interface looks like public interface SenseiMapReduce<MapResult extends Serializable, ReduceResult extends Serializable> extends Serializable, but Map is not itself serializable. Perhaps that's something we can rethink.

Comment (chtyim) [inline sensei-core/src/test/java/com/senseidb/search/req/mapred/FacetCountsMapReduce.java:52]: It seems like a bad good design for public interface to extends from Serializable. We should rather have a Serializer interface that can serialize SenseiMapReduce, which also allow us using different serialization strategy.

Comment (vzhabiuk) [inline sensei-core/src/test/java/com/senseidb/search/req/mapred/FacetCountsMapReduce.java:52]: A good point. That's something that we can improve

2012/10/4 Terence Yim notifications@github.com

**Pull request merged**